subserver_mgr.go

```diff
@@ -0,0 +1,399 @@
+			err := s.remoteConn.Close()
+			if err != nil {
+				returnErr = fmt.Errorf("could not close "+
+					"remote connection: %v", err)
+			}
+			return
+		}
+
+		// Else, stop the integrated sub-server process.
+		err := s.cfg.stopIntegrated()
+		if err != nil {
+			returnErr = fmt.Errorf("could not close "+
+				"integrated connection: %v", err)
+			return
+		}
+
+		if s.cfg.serverErrChan == nil {
+			return
+		}
+
+		returnErr = <-s.cfg.serverErrChan
+	})
+
+	return returnErr
+}
+
+// startIntegrated starts the subServer in integrated mode.
+func (s *subServer) startIntegrated(lndClient lnrpc.LightningClient,
+	lndGrpc *lndclient.GrpcLndServices, withMacaroonService bool) {
+
+	err := s.cfg.startIntegrated(lndClient, lndGrpc, withMacaroonService)
+	if err != nil {
+		s.cfg.onStartError(err)
+		return
+	}
+	s.integratedStarted = true
+	s.cfg.onStartSuccess()
+
+	if s.cfg.serverErrChan == nil {
+		return
+	}
+
+	go func() {
+		select {
+		case err := <-s.cfg.serverErrChan:
+			// The sub server should shut itself down if an error
+			// happens. We don't need to try to stop it again.
+			s.integratedStarted = false
+
+			s.cfg.onStartError(fmt.Errorf("received "+
+				"critical error from sub-server, "+
+				"shutting down: %v", err),
+			)
+
+		case <-s.quit:
+		}
+	}()
+}
+
+// connectRemote attempts to make a connection to the remote sub-server.
+func (s *subServer) connectRemote() {
+	var err error
+	s.remoteConn, err = dialBackend(
+		s.cfg.name, s.cfg.remoteCfg.RPCServer,
+		lncfg.CleanAndExpandPath(s.cfg.remoteCfg.TLSCertPath),
+	)
+	if err != nil {
+		err := fmt.Errorf("remote dial error: %v", err)
+		s.cfg.onStartError(err)
+		return
+	}
+	s.cfg.onStartSuccess()
+}
+
+// newSubServer constructs a new subServer using the given config.
+func newSubServer(cfg *subServerCfg) *subServer {
+	return &subServer{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}
+}
+
+// subServerMgr manages a set of subServer objects.
+type subServerMgr struct {
+	servers []*subServer
+	mu      sync.RWMutex
+}
+
+// newSubServerMgr constructs a new subServerMgr.
+func newSubServerMgr() *subServerMgr {
+	return &subServerMgr{
+		servers: []*subServer{},
+	}
+}
+
+// AddServer adds a new subServer to the manager's set.
+func (s *subServerMgr) AddServer(server *subServer) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.servers = append(s.servers, server)
+}
+
+// StartIntegratedServers starts all the manager's sub-servers that should be
+// started in integrated mode.
+func (s *subServerMgr) StartIntegratedServers(lndClient lnrpc.LightningClient,
+	lndGrpc *lndclient.GrpcLndServices, withMacaroonService bool) {
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, ss := range s.servers {
+		if ss.cfg.remote {
+			continue
+		}
+
+		ss.startIntegrated(lndClient, lndGrpc, withMacaroonService)
+	}
+}
+
+// StartRemoteSubServers creates connections to all the manager's sub-servers
+// that are running remotely.
+func (s *subServerMgr) StartRemoteSubServers() {
```

Comment:
The func name can be more accurate if we use something like `ConnectRemoteSubServers` given that we are not starting anything.